coresimd/src/x86/i686/aes.rs

```diff
@@ -0,0 +1,144 @@
+use x86::__m128i;
+
+#[cfg(test)]
+use stdsimd_test::assert_instr;
+
+#[allow(improper_ctypes)]
+extern "C" {
+    #[link_name = "llvm.x86.aesni.aesdec"]
+    fn aesdec(a: __m128i, round_key: __m128i) -> __m128i;
+    #[link_name = "llvm.x86.aesni.aesdeclast"]
+    fn aesdeclast(a: __m128i, round_key: __m128i) -> __m128i;
+    #[link_name = "llvm.x86.aesni.aesenc"]
+    fn aesenc(a: __m128i, round_key: __m128i) -> __m128i;
+    #[link_name = "llvm.x86.aesni.aesenclast"]
+    fn aesenclast(a: __m128i, round_key: __m128i) -> __m128i;
+    #[link_name = "llvm.x86.aesni.aesimc"]
+    fn aesimc(a: __m128i) -> __m128i;
+    #[link_name = "llvm.x86.aesni.aeskeygenassist"]
+    fn aeskeygenassist(a: __m128i, imm8: u8) -> __m128i;
+}
+
+/// Perform one round of an AES decryption flow on data (state) in `a`.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aesdec))]
+pub unsafe fn _mm_aesdec_si128(a: __m128i, round_key: __m128i) -> __m128i {
+    aesdec(a, round_key)
+}
+
+/// Perform the last round of an AES decryption flow on data (state) in `a`.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aesdeclast))]
+pub unsafe fn _mm_aesdeclast_si128(a: __m128i, round_key: __m128i) -> __m128i {
+    aesdeclast(a, round_key)
+}
+
+/// Perform one round of an AES encryption flow on data (state) in `a`.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aesenc))]
+pub unsafe fn _mm_aesenc_si128(a: __m128i, round_key: __m128i) -> __m128i {
+    aesenc(a, round_key)
+}
+
+/// Perform the last round of an AES encryption flow on data (state) in `a`.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aesenclast))]
+pub unsafe fn _mm_aesenclast_si128(a: __m128i, round_key: __m128i) -> __m128i {
+    aesenclast(a, round_key)
+}
+
+/// Perform the “InvMixColumns” transformation on `a`.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aesimc))]
+pub unsafe fn _mm_aesimc_si128(a: __m128i) -> __m128i {
+    aesimc(a)
+}
+
+/// Assist in expanding the AES cipher key.
+///
+/// Assist in expanding the AES cipher key by computing steps towards generating
+/// a round key for encryption cipher using data from `a` and an 8-bit round constant.
+#[inline]
+#[target_feature(enable = "aes")]
+#[cfg_attr(test, assert_instr(aeskeygenassist))]
```

Comment:
Change this

```rust
assert_instr(aeskeygenassist)
```

to (from the top of my head):

```rust
assert_instr(aeskeygenassist, imm8 = 0)
```

That should fix that.

Reply:
Just to explain what's going on, `imm8` must be a compile-time argument, and while testing the disassembly, we need to pass it a value. If one doesn't specify it in the `assert_instr` macro, it will pass it a run-time value, and that won't work for this intrinsic. 

We don't have a way of expressing that function arguments must be const yet.

Reply:
I see, thanks for the explanation. Should be fixed now.

I was a bit worried that this meant that the entire match statement would end up in the final binary, so I also verified the disassembly of `test_mm_aeskeygenassist_si128` in release mode, and fortunately it does get const-folded:

    0000000000041720 <coresimd::x86::i686::aes::tests::test_mm_aeskeygenassist_si128::test_mm_aeskeygenassist_si128>:
       41720:       55                      push   %rbp
       41721:       48 89 e5                mov    %rsp,%rbp
       41724:       48 81 ec 80 00 00 00    sub    $0x80,%rsp
       4172b:       66 0f 3a df 05 3b c7    aeskeygenassist $0x5,0x11c73b(%rip),%xmm0        # 15de70 <str.4y+0x20>
       41732:       11 00 05 
       41735:       66 0f 7f 45 90          movdqa %xmm0,-0x70(%rbp)
       4173a:       0f 28 05 9f c7 11 00    movaps 0x11c79f(%rip),%xmm0        # 15dee0 <str.4y+0x90>
       41741:       0f 29 45 80             movaps %xmm0,-0x80(%rbp)

In debug mode it doesn’t though, instead there is a call to `_mm_aeskeygenassist_si128` which contains the entire match statement. Something along the lines of rust-lang/rust#47980 would be good to have indeed.